### PR TITLE
Move MPIStateArray reductions to the device

### PR DESCRIPTION
**Pull request by lcw**
This uses LazyArrays to build up expressions that are evaluated in `mapreduce` on the compute device.  This minimizes the data transfer between the device and the host for reductions.  This is useful for the linear solve in IMEX based time stepping.  This will also be useful for efficient calculation of state array statistics.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Should this get a test?

**Comment by lcw** [comment]
Good question.  I didn't add any new functions, just moved part of the reduction to use `mapreduce` so it is being tested as much as it was before.  Are there additional tests you want me to add?

I did add a test to the broadcasted GPU reduction in https://github.com/JuliaGPU/GPUArrays.jl/pull/196.

File: src/Arrays/MPIStateArrays.jl [hunk 1]
```diff
@@ -353,39 +300,23 @@ the listed states are summed, otherwise all the states in `A` are used.
...
-  @inbounds for e = elems
-    for q = states, i = 1:Np
-      wsum += weights[i, e] * A[i, q, e]
-    end
-  end
+  locwsum = mapreduce(identity, +, E, init=init)
 
-  wsum
+  # Need to use anomous function version of sum else MPI.jl using MPI_SUM
+  T(MPI.Allreduce([locwsum], (x,y)->x+y, A.mpicomm)[1])
```
Comment (simonbyrne) [inline src/Arrays/MPIStateArrays.jl]: why can't you use `MPI_SUM`?

**Comment by simonbyrne** [comment]
I didn't realise MPIStateArrays didn't have any of their own tests. That would be good to have, but we can do those in another PR.

File: src/Arrays/MPIStateArrays.jl (hunk 1 above)
Comment (lcw) [reply src/Arrays/MPIStateArrays.jl]: That was there before so I kept it.  I believe `MPI.jl` was selecting `MPI_SUM` for variables of type `DoubleFloat{DFloat}` which doesn't work.  However, I didn't test it.

**Comment by lcw** [comment]
The only test failure I get is for
```
mpirun -n 1 julia --project=env/gpu test/DGmethods/compressible_Navier_Stokes/mms_bc.jl
```
which happens when I update the packages (aka dc9c495).  I tried updating all of the packages and still see the same issue.  It seems something that used to work on the GPU now doesn't.  I am not exactly sure what is happening. @vchuravy, is the following stack trace at all enlightening?

```
❯ mpirun -n 1 julia --project=env/gpu test/DGmethods/compressible_Navier_Stokes/mms_bc.jl
[ Info: (CuArray, Float64, 2)
┌ Info: Starting
└ norm(Q₀) = 1.4718488046151097e+02
/home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl: Error During Test at /home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl:331
  Got exception outside of a @test
  InvalidIRError: compiling volumeviscterms!(Cassette.Context{nametype(Ctx),Nothing,Nothing,getfield(GPUifyLoops, Symbol("##PassType#371")),Nothing,Cassette.DisableHooks}, typeof(CLIMA.DGBalanceLawDiscretizations.volumeviscterms!), Val{2}, Val{4}, Val{5}, Val{(1, 2, 3, 4)}, Val{3}, Val{6}, Val{3}, typeof(compute_stresses!), typeof(velocities!), CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CuDeviceArray{Float64,3,CUDAnative.AS.Global}, CuDeviceArray{Float64,3,CUDAnative.A
S.Global}, CuDeviceArray{Float64,3,CUDAnative.AS.Global}, Float64, CuDeviceArray{Float64,2,CUDAnative.AS.Global}, UnitRange{Int64}) resulted in invalid LLVM IR
  Reason: unsupported dynamic function invocation (call to _getindex(a::StaticArray, ::Length{L}, ::Colon) where L in StaticArrays at /home/lucas/.julia/packages/StaticArrays/VyRz3/src/indexing.jl:92)
  Stacktrace:
   [1] call at /home/lucas/.julia/packages/Cassette/xggAf/src/context.jl:447
   [2] fallback at /home/lucas/.julia/packages/Cassette/xggAf/src/context.jl:445
   [3] _overdub_fallback at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:465
   [4] getindex at /home/lucas/.julia/packages/StaticArrays/VyRz3/src/indexing.jl:88
   [5] getindex at /home/lucas/.julia/packages/GPUifyLoops/HaVjN/src/scratch.jl:45
   [6] volumeviscterms! at /home/lucas/research/code/CLIMA/src/DGmethods/DGBalanceLawDiscretizations_kernels.jl:291
   [7] overdub at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:0
  Reason: unsupported dynamic function invocation (call to overdub)
  Stacktrace:
   [1] volumeviscterms! at /home/lucas/research/code/CLIMA/src/DGmethods/DGBalanceLawDiscretizations_kernels.jl:291
   [2] overdub at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:0
  Reason: unsupported dynamic function invocation (call to _getindex(a::StaticArray, ::Length{L}, ::Colon) where L in StaticArrays at /home/lucas/.julia/packages/StaticArrays/VyRz3/src/indexing.jl:92)
  Stacktrace:
   [1] call at /home/lucas/.julia/packages/Cassette/xggAf/src/context.jl:447
   [2] fallback at /home/lucas/.julia/packages/Cassette/xggAf/src/context.jl:445
   [3] _overdub_fallback at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:465
   [4] getindex at /home/lucas/.julia/packages/StaticArrays/VyRz3/src/indexing.jl:88
   [5] getindex at /home/lucas/.julia/packages/GPUifyLoops/HaVjN/src/scratch.jl:45
   [6] volumeviscterms! at /home/lucas/research/code/CLIMA/src/DGmethods/DGBalanceLawDiscretizations_kernels.jl:325
   [7] overdub at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:0
  Reason: unsupported dynamic function invocation (call to overdub)
  Stacktrace:
   [1] volumeviscterms! at /home/lucas/research/code/CLIMA/src/DGmethods/DGBalanceLawDiscretizations_kernels.jl:325
   [2] overdub at /home/lucas/.julia/packages/Cassette/xggAf/src/overdub.jl:0
  Stacktrace:
   [1] check_ir(::CUDAnative.CompilerJob, ::LLVM.Module) at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/validation.jl:114
   [2] macro expansion at /home/lucas/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
   [3] #codegen#121(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.codegen), ::Symbol, ::CUDAnative.CompilerJob) at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/driver.jl:186
   [4] #codegen at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/driver.jl:0 [inlined]
   [5] #compile#120(::Bool, ::Bool, ::Bool, ::Bool, ::Bool, ::typeof(CUDAnative.compile), ::Symbol, ::CUDAnative.CompilerJob) at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/driver.jl:47
   [6] #compile at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/common.jl:0 [inlined]
   [7] #compile#119 at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/compiler/driver.jl:28 [inlined]
   [8] #compile at ./none:0 [inlined] (repeats 2 times)
   [9] macro expansion at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/execution.jl:388 [inlined]
   [10] #cufunction#161(::String, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(cufunction), ::typeof(Cassette.overdub), ::Type{Tuple{Cassette.Context{nametype(Ctx),Nothing,Nothing,getfield(GPUifyLoops, Symbol("##PassType#371")),Nothing,Cassette.DisableHooks},typeof(CLIMA.DGBalanceLawDiscretizations.volumeviscterms!),Val{2},Val{4},Val{5},Val{(1, 2, 3, 4)},Val{3},Val{6},Val{3},typeof(compute_stresses!),typeof(velocities!),CuDeviceArray{Flo
at64,3,CUDAnative.AS.Global},CuDeviceArray{Float64,3,CUDAnative.AS.Global},CuDeviceArray{Float64,3,CUDAnative.AS.Global},CuDeviceArray{Float64,3,CUDAnative.AS.Global},Float64,CuDeviceArray{Float64,2,CUDAnative.AS.Global},UnitRange{Int64}}}) at /home/lucas/.julia/packages/CUDAnative/q7ezN/src/execution.jl:356
   [11] (::getfield(CUDAnative, Symbol("#kw##cufunction")))(::NamedTuple{(:name,),Tuple{String}}, ::typeof(cufunction), ::Function, ::Type) at ./none:0
   [12] macro expansion at /home/lucas/.julia/packages/GPUifyLoops/HaVjN/src/GPUifyLoops.jl:125 [inlined]
   [13] macro expansion at ./gcutils.jl:87 [inlined]
   [14] #launch#46(::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol},NamedTuple{(:threads, :blocks),Tuple{Tuple{Int64,Int64,Int64},Int64}}}, ::typeof(GPUifyLoops.launch), ::GPUifyLoops.CUDA, ::typeof(CLIMA.DGBalanceLawDiscretizations.volumeviscterms!), ::Val{2}, ::Vararg{Any,N} where N) at /home/lucas/.julia/packages/GPUifyLoops/HaVjN/src/GPUifyLoops.jl:121
   [15] (::getfield(GPUifyLoops, Symbol("#kw##launch")))(::NamedTuple{(:threads, :blocks),Tuple{Tuple{Int64,Int64,Int64},Int64}}, ::typeof(GPUifyLoops.launch), ::GPUifyLoops.CUDA, ::typeof(CLIMA.DGBalanceLawDiscretizations.volumeviscterms!), ::Val{2}, ::Vararg{Any,N} where N) at ./none:0
   [16] macro expansion at /home/lucas/.julia/packages/GPUifyLoops/HaVjN/src/GPUifyLoops.jl:54 [inlined]
   [17] #odefun!#13(::Bool, ::typeof(CLIMA.SpaceMethods.odefun!), ::DGBalanceLaw, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::Float64) at /home/lucas/research/code/CLIMA/src/DGmethods/DGBalanceLawDiscretizations.jl:610
   [18] (::getfield(CLIMA.SpaceMethods, Symbol("#kw##odefun!")))(::NamedTuple{(:increment,),Tuple{Bool}}, ::typeof(CLIMA.SpaceMethods.odefun!), ::DGBalanceLaw, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::Float64) at ./none:0
   [19] (::getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("###12#13#15")){DGBalanceLaw})(::Bool, ::getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("##12#14")){getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("###12#13#15")){DGBalanceLaw}}, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::Vararg{Any,N} where N) at /home/lucas/research/code/CLIMA/src/ODESolvers/LowStorageRungeKuttaMethod.jl:68
   [20] (::getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("#kw###12#14")))(::NamedTuple{(:increment,),Tuple{Bool}}, ::getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("##12#14")){getfield(CLIMA.LowStorageRungeKuttaMethod, Symbol("###12#13#15")){DGBalanceLaw}}, ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::Vararg{Any,N} where N) at ./none:0
   [21] dostep!(::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::LowStorageRungeKutta2N{Float64,Float64,MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}},5}, ::Int64, ::Bool) at /home/lucas/research/code/CLIMA/src/ODESolvers/LowStorageRungeKuttaMethod.jl:91
   [22] #solve!#9(::Int64, ::Bool, ::Int64, ::Tuple{CLIMA.GenericCallbacks.EveryXWallTimeSeconds,CLIMA.GenericCallbacks.EveryXSimulationSteps}, ::typeof(solve!), ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::LowStorageRungeKutta2N{Float64,Float64,MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}},5}) at /home/lucas/research/code/CLIMA/src/ODESolvers/ODESolvers.jl:70
   [23] (::getfield(CLIMA.ODESolvers, Symbol("#kw##solve!")))(::NamedTuple{(:timeend, :callbacks),Tuple{Int64,Tuple{CLIMA.GenericCallbacks.EveryXWallTimeSeconds,CLIMA.GenericCallbacks.EveryXSimulationSteps}}}, ::typeof(solve!), ::MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}}, ::LowStorageRungeKutta2N{Float64,Float64,MPIStateArray{Tuple{25,5},Float64,CuArray,3,CuArray{Float64,3},2,CuArray{Int64,1}},5}) at ./none:0
   [24] run(::MPI.Comm, ::Type, ::Int64, ::BrickTopology{2,Float64}, ::Function, ::Int64, ::Int64, ::Type, ::Float64) at /home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl:285
   [25] top-level scope at /home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl:366
   [26] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1186
   [27] top-level scope at /home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl:331
   [28] include at ./boot.jl:328 [inlined]
   [29] include_relative(::Module, ::String) at ./loading.jl:1094
   [30] include(::Module, ::String) at ./Base.jl:31
   [31] exec_options(::Base.JLOptions) at ./client.jl:295
   [32] _start() at ./client.jl:464

Test Summary:                                                                       | Error  Total
/home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl |     1      1
ERROR: LoadError: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /home/lucas/research/code/CLIMA/test/DGmethods/compressible_Navier_Stokes/mms_bc.jl:306
```

**Comment by lcw** [comment]
bors try

File: src/Arrays/MPIStateArrays.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Arrays/MPIStateArrays.jl]: Ah right

**Comment by lcw** [comment]
bors try

**Comment by vchuravy** [comment]
Looking into it

**Comment by vchuravy** [comment]
Narrowed it down to:
```
using StaticArrays
using GPUifyLoops
f() = StaticArrays._Length(1)
f()
# f() = Length(1)
GPUifyLoops.Cassette.overdub(GPUifyLoops.ctx, f)
# Length(StaticArrays.Dynamic())
```

Working on a fix.

**Comment by glwagner** [comment]
@ali-ramadhan this could be related to the issue we are having with function composition.

**Comment by lcw** [comment]
bors try

File: src/Arrays/MPIStateArrays.jl [hunk 2]
```diff
@@ -258,91 +258,38 @@ end
...
-  @inbounds for e = elems, q = 1:nstate, i = 1:Np
-    nrm += weights[i, e] * Q[i, q, e]^2
+  if ~isempty(Q.weights)
+    w = @view Q.weights[:, :, Q.realelems]
+    E = @~ E .* w
   end
 
-  nrm
+  locnorm = mapreduce(identity, op, E, init=init)
+  sqrt(MPI.Allreduce([locnorm], mpiop, Q.mpicomm)[1])
```
Comment (mwarusz) [inline src/Arrays/MPIStateArrays.jl]: This sqrt is only correct for p == 2, if you want this function to work for general p you need x ^ (1 / p) and handle the case of p == Inf.

Comment (lcw) [reply src/Arrays/MPIStateArrays.jl]: Good catch, thanks!

**Comment by lcw** [comment]
@simonbyrne  I tried to update all of the packages and it seems to break things on Julia v1 and v1.1.  It seems to be working on nightly (and v1.2rc1 locally).  I am not sure what todo.

**Comment by simonbyrne** [comment]
Not sure what happened: I wasn't able to do `Pkg.resolve()` on 1.1, so I just reset it.

**Comment by lcw** [comment]
I now tried Julia v1.1.0 to do the update

**Comment by lcw** [comment]
I am sorry if my force push killed what you did

**Comment by simonbyrne** [comment]
That's fine, it was exactly the same thing.

**Comment by lcw** [comment]
bors try

**Comment by lcw** [comment]
Looks like if we update to the latest GPUifyLoops we need to drop support for v1 of Julia.

**Comment by simonbyrne** [comment]
That's fine by me: as long as we're running on at least one release version of Julia, I'm happy. Feel free to edit Travis/Appveyor files as needed.

**Comment by lcw** [comment]
bors try

**Comment by lcw** [comment]
bors try

**Comment by simonbyrne** [comment]
okay, it passed that time (before the thread limit changes). Not sure what is going on.

**Comment by simonbyrne** [comment]
(gpu-tests are still timing out, but #282 will fix that).

**Pull request merged**